src/rt/lifetime.d

```diff
@@ -1193,24 +1186,52 @@ extern (C) void _d_delarray_t(void[]* p, const TypeInfo_Struct ti)
 {
     if (p)
     {
-        assert(!(*p).length || (*p).ptr);
-        if ((*p).ptr)
+        auto bic = __getBlkInfo(p.ptr);
+        auto info = bic ? *bic : GC.query(p.ptr);
+
+        if (info.base && (info.attr & BlkAttr.APPENDABLE))
         {
             if (ti) // ti non-null only if ti is a struct with dtor
-                finalize_array(p.ptr, p.length * ti.tsize, ti);
-
-            // if p is in the cache, clear it as well
-            if(auto bic = __getBlkInfo((*p).ptr))
             {
-                // clear the data from the cache, it's being deleted.
-                bic.base = null;
+                void* start = __arrayStart(info);
+                size_t length = __arrayAllocLength(info, ti);
+                finalize_array(start, length, ti);
             }
-            GC.free((*p).ptr);
+
+            // if p is in the cache, clear it there as well
+            if(bic)
+                bic.base = null;
+
+            GC.free(info.base);
         }
         *p = null;
```

Comment:
Not changing p if it's not freed sounds reasonable. Would that be a change to the language specification? Unfortunately the documentation is pretty short for "delete".